src/unity/python/turicreate/toolkits/_mps_utils.py

```diff
@@ -244,6 +244,50 @@ def _xavier_init(weight):
     return _np.random.uniform(-c, c, shape).astype(_np.float32)
 
 
+class MpsFloatArray(object):
+    """
+    A Python wrapper owning a C++ float_array created by the TCMPS backend.
+
+    This class exists to simplify conversions from numpy to the TCMPS format and
+    to simplify memory management. Instances usually just serve as arguments to
+    the methods on MpsGraphAPI and MpsLowLevelAPI, below.
+    """
+
+    def __init__(self, x):
+        """Wrap a numpy array"""
+
+        # Load TCMPS backend library.
+        self._LIB = _load_tcmps_lib()
+        assert self._LIB is not None, "Cannot use MpsFloatArray without libtcmps.dylib"
+
+        # Convert the input if necessary to contain a contiguous float array.
+        self.data = x
+        if self.data.dtype != _np.float32:
+            self.data = self.data.astype(_np.float32)
+        if not self.data.flags.c_contiguous:
+            self.data = self.data.copy()
+        assert self.data.flags.c_contiguous, "Data must be row-major"
+
+        # Obtain a pointer to the internal float array (and obtain size).
+        data_ptr = self.data.ctypes.data_as(_ctypes.POINTER(_ctypes.c_void_p))
+        sz = _ctypes.c_size_t(self.data.size)
+
+        # Copy the shape so that it contains a size_t array.
+        self.shape = _np.array(self.data.shape).astype(_np.uintp)
+        shape_ptr = self.shape.ctypes.data_as(_ctypes.POINTER(_ctypes.c_size_t))
+        dim = _ctypes.c_size_t(self.data.ndim)
+
+        # Call into TCMPS to create a wrapper around self.data and self.shape.
+        # Those two properties must outlive the resulting self.handle.
+        self.handle = _ctypes.c_void_p()
+        status_code = self._LIB.TCMPSCreateFloatArray(
+            _ctypes.byref(self.handle), data_ptr, sz, shape_ptr, dim)
+        assert status_code == 0, "Error calling TCMPSCreateFloatArray"
+
+    def __del__(self):
+        status_code = self._LIB.TCMPSDeleteFloatArray(self.handle)
```

Comment:
This looks like it's freeing memory. If so, I don't think `__del__` is the right place for that. From [the documentation](https://docs.python.org/3.6/reference/datamodel.html#object.__del__) for `__del__`: It is not guaranteed that `__del__()` methods are called for objects that still exist when the interpreter exits.

Reply:
By the time the interpreter exits, practically it doesn't matter anymore since the OS will clean up any unfreed memory. I agree it's bad practice to rely on this, but it's unclear what a better place is for this though if that's the behavior of `__del__`. It seems like the alternative is manual management (which can already be made through `delete` if one pleases).

Reply:
It is probably fine. I think generally context managers are the preferred method for doing operation which require you to free memory, but I don't know if it would be feasible to use that instead here.

Reply:
Yeah, AFAICT the semantics for `__del__` are the same as those for a C++ destructor, and it's quite possible for a C++ program to end with several objects on the heap whose destructors are never called. Generally, this should only be a problem if the object controls state that can persist past the process's lifetime, such as a file.

Yeah, using a context manager would definitely be a more Pythonic way of doing this kind of thing, although it's not clear to me how that works for things like these float arrays that are being returned from some C function. Honestly, I'm hoping all this code will go away someday once we push the TCMPS client code into C++; for now, this is just another instance of a pattern (of using __del__ to destroy C++ objects) that we already shipped in TC5.0 (see the `MpsGraphApi` class in the same file).

Reply:
Sounds good.